### ISSUE TITLE
Discover pattern(s) within cloning git repos from chef attributes

### ISSUE THREAD

**Issue by informaticianme** [informaticianme/chef.cookbooks#22]
Relates to #20, #24

Currently, I use some of the attributes from users.rb to clone a bunch of repositories within the users.rb recipe. I also use attributes from rbenv.rb to clone repositories in the rbenv.rb recipe. The methods I use to achieve these clonings are very different but the content (if not the shape of it) is very similar.

Some questions to consider:
* Should the attributes be represented in a flat (rbenv.rb) structure? or a hierarchical (users.rb) one?
* How much (and which) data can or should be pulled from recipes and placed in attributes (rbenv.rb, users.rb)?

**Comment by informaticianme** [comment]
Part of the complexity regarding the second question is the seeking to understand how variables (attributes) get processed between role, node, or environment attributes, cookbook attributes, and then recipes (it's more complex but those are the basics).

On the topic of attribute processing, [coderanger] has written a great post about the use of attributes within attributes: https://coderanger.net/derived-attributes/

**Comment by informaticianme** [comment]
The major issue that kept me hung up when considering this issue before was that I wanted to move lots of data from the recipe files into the attributes files but would hit a single common snag with wanting to move the `"/home/#{user}"` portion of the git destination path from the recipe to the attribute but couldn't see a way to do it because it was dependent upon the value of `#{user}` within the recipe itself. Using `lazy` evaluation or `.run_action` may be a viable way around this.

Explore `lazy` / `.run_action()` as a possible viable method for moving more data from recipes into attributes. This would help to build the pattern sought in the original description of this issue.

**Comment by informaticianme** [comment]
I ended up doing a deep exploration into this topic. The basic questions were:
1. [x] Should I try to combine all git repos into one area of the `users.rb` json file, or should some be separate?
There are both user dependent and user independent git repos. Only track user dependent repos within the `attributes/user.rb`.
1. [x] Should git repo info always be hierarchical or flat, or is context dependent?
It is context dependent. The `attributes/users.rb` should use a hierarchical format as there are very many to track.
1. [x] Should I try to create a wrapper for the git LWRP?
No. A LWRP wrapper would be overkill for this issue. The current one works fine.
1. [x] Should I try to create a common function for git blocks?
I explored this and determined that a common function *would* be appropriate except that writing a function for both git repos that require ssh key access and ones that do not is not a clean way to code. I also determined that although there is some duplication of properties, it seems best to have the properties listed explicitly. Also, much of the wetness of the code has been refactored away without need of a common git function.
